curvefs/src/client/xattr_manager.cpp

```diff
@@ -0,0 +1,543 @@
+                        }
+                        summaryInfo->entries++;
+                        summaryInfo->fbytes += it.length();
+                        // record hardlink
+                        if (it.type() != FsFileType::TYPE_DIRECTORY &&
+                            it.nlink() > 1) {
+                            std::lock_guard<std::mutex> guard(*mapMutex);
+                            auto iter = hardLinkMap->find(it.inodeid());
+                            if (iter != hardLinkMap->end()) {
+                                iter->second += it.length();
+                            } else {
+                                hardLinkMap->emplace(it.inodeid(), 0);
+                            }
+                        }
+                    }
+                } else {
+                    ret->store(false);
+                    break;
+                }
+            }
+        }
+    }
+}
+
+CURVEFS_ERROR XattrManager::CalAllLayerSumInfo(Inode *inode) {
+    std::stack<uint64_t> iStack;
+    std::mutex stackMutex;
+
+    // record hard link, <inodeId, need2minus>
+    std::unordered_map<uint64_t, uint64_t> hardLinkMap;
+    std::mutex mapMutex;
+
+    SummaryInfo summaryInfo;
+    std::mutex valueMutex;
+
+    auto ino = inode->inodeid();
+    iStack.emplace(ino);
+    std::vector<Thread> threadpool;
+    Atomic<uint32_t> inflightNum(0);
+    Atomic<bool> ret(true);
+
+    for (auto i = listDentryThreads_; i > 0; i--) {
+        threadpool.emplace_back(Thread(
+                        &XattrManager::ConcurrentGetInodeAttr,
+                        this, &iStack, &stackMutex, &hardLinkMap, &mapMutex,
+                        &summaryInfo, &valueMutex, &inflightNum, &ret));
+    }
+
+    for (auto &thread : threadpool) {
+        thread.join();
+    }
+
+    if (!ret.load()) {
+        return CURVEFS_ERROR::INTERNAL;
+    }
+
+    // deal with hardlink
+    for (const auto &it : hardLinkMap) {
+        summaryInfo.fbytes -= it.second;
+    }
+
+    inode->mutable_xattr()->insert({XATTRRFILES,
+        std::to_string(summaryInfo.files)});
+    inode->mutable_xattr()->insert({XATTRRSUBDIRS,
+        std::to_string(summaryInfo.subdirs)});
+    inode->mutable_xattr()->insert({XATTRRENTRIES,
+        std::to_string(summaryInfo.entries)});
+    inode->mutable_xattr()->insert({XATTRRFBYTES,
+        std::to_string(summaryInfo.fbytes + inode->length())});
+    return CURVEFS_ERROR::OK;
+}
+
+void XattrManager::ConcurrentGetInodeXattr(
+    std::stack<uint64_t> *iStack,
+    std::mutex *stackMutex,
+    Inode *inode,
+    std::mutex *inodeMutex,
+    Atomic<uint32_t> *inflightNum,
+    Atomic<bool> *ret) {
+    while (1) {
+        std::list<Dentry> dentryList;
+        std::set<uint64_t> inodeIds;
+        std::list<XAttr> xattrs;
+        auto tret = ConcurrentListDentry(&dentryList, iStack, stackMutex,
+                                         true, inflightNum, ret);
+        if (!tret) {
+            break;
+        } else {
```

Comment:
You can remove the `else` block, it can make logic more clearly:
```
if (!tret) {
    break;
}

// do something for else
```

Reply:
fixed all